nexus/db-queries/src/db/pool.rs

```diff
@@ -3,108 +3,144 @@
-//
 // TODO-design Need TLS support (the types below hardcode NoTls).
 
 use super::Config as DbConfig;
-use async_bb8_diesel::ConnectionError;
-use async_bb8_diesel::ConnectionManager;
+use crate::db::pool_connection::DieselPgConnector;
+
+use qorb::backend;
+use qorb::policy::Policy;
+use qorb::resolver::{AllBackends, Resolver};
+use qorb::resolvers::dns::{DnsResolver, DnsResolverConfig};
+use qorb::service;
+use std::collections::BTreeMap;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::sync::watch;
 
 pub use super::pool_connection::DbConnection;
 
+type QorbConnection = async_bb8_diesel::Connection<DbConnection>;
+type QorbPool = qorb::pool::Pool<QorbConnection>;
+
 /// Wrapper around a database connection pool.
 ///
 /// Expected to be used as the primary interface to the database.
 pub struct Pool {
-    pool: bb8::Pool<ConnectionManager<DbConnection>>,
+    inner: QorbPool,
 }
 
-impl Pool {
-    pub fn new(log: &slog::Logger, db_config: &DbConfig) -> Self {
-        // Make sure diesel-dtrace's USDT probes are enabled.
-        usdt::register_probes().expect("Failed to register USDT DTrace probes");
-        Self::new_builder(log, db_config, bb8::Builder::new())
-    }
+// Provides an alternative to the DNS resolver for cases where we want to
+// contact the pool directly.
+struct SingleHostResolver {
+    tx: watch::Sender<AllBackends>,
+}
 
-    pub fn new_failfast_for_tests(
-        log: &slog::Logger,
-        db_config: &DbConfig,
-    ) -> Self {
-        Self::new_builder(
-            log,
-            db_config,
-            bb8::Builder::new()
-                .connection_timeout(std::time::Duration::from_millis(1)),
-        )
+impl SingleHostResolver {
+    fn new(config: &DbConfig) -> Self {
+        let backends = Arc::new(BTreeMap::from([(
+            backend::Name::new("singleton"),
+            backend::Backend { address: config.url.address() },
+        )]));
+        let (tx, _rx) = watch::channel(backends.clone());
+        Self { tx }
     }
+}
 
-    fn new_builder(
-        log: &slog::Logger,
-        db_config: &DbConfig,
-        builder: bb8::Builder<ConnectionManager<DbConnection>>,
-    ) -> Self {
-        let url = db_config.url.url();
-        let log = log.new(o!(
-            "database_url" => url.clone(),
-            "component" => "db::Pool"
-        ));
-        info!(&log, "database connection pool");
-        let error_sink = LoggingErrorSink::new(log);
-        let manager = ConnectionManager::<DbConnection>::new(url);
-        let pool = builder
-            .connection_customizer(Box::new(
-                super::pool_connection::ConnectionCustomizer::new(),
-            ))
-            .error_sink(Box::new(error_sink))
-            .build_unchecked(manager);
-        Pool { pool }
+impl Resolver for SingleHostResolver {
+    fn monitor(&mut self) -> watch::Receiver<AllBackends> {
+        self.tx.subscribe()
     }
+}
```

Comment:
not a blocker or anything, but: it occurs to me that `qorb` could provide an `impl Resolver for watch::Receiver<AllBackends>` that just clones the rx, which would kind of obviate the need for this type..

Reply:
Hrmmm I suppose so, but the choice of resolver still needs to be caller-controlled. I think we could probably adjust the API of `Pool::new` to take a `watch::Receiver<AllBackends>` instead of a Boxed `Resolver`, but they feel pretty similar functionally, right?